aggregator_core/src/datastore.rs

```diff
@@ -4721,6 +4722,111 @@ impl<C: Clock> Transaction<'_, C> {
             .await?;
         check_single_row_mutation(self.execute(&stmt, &[&aggregator_url, &role]).await?)
     }
+
+    /// Get the [`TaskUploadCounter`] for a task. This is aggregated across all shards.
+    #[tracing::instrument(skip(self), err)]
+    pub async fn get_task_upload_counter(
+        &self,
+        task_id: &TaskId,
+    ) -> Result<TaskUploadCounter, Error> {
+        let stmt = self
+            .prepare_cached(
+                "SELECT
+                    COALESCE(SUM(interval_collected)::BIGINT, 0) AS interval_collected,
+                    COALESCE(SUM(report_decode_failure)::BIGINT, 0) AS report_decode_failure,
+                    COALESCE(SUM(report_decrypt_failure)::BIGINT, 0) AS report_decrypt_failure,
+                    COALESCE(SUM(report_expired)::BIGINT, 0) AS report_expired,
+                    COALESCE(SUM(report_outdated_key)::BIGINT, 0) AS report_outdated_key,
+                    COALESCE(SUM(report_success)::BIGINT, 0) AS report_success,
+                    COALESCE(SUM(report_too_early)::BIGINT, 0) AS report_too_early,
+                    COALESCE(SUM(task_expired)::BIGINT, 0) AS task_expired
+                FROM task_upload_counters
+                WHERE task_id = (SELECT id FROM tasks WHERE task_id = $1)",
+            )
+            .await?;
+
+        let row = self.query_one(&stmt, &[task_id.as_ref()]).await?;
+        Ok(TaskUploadCounter {
+            task_id: *task_id,
+            interval_collected: row.get_bigint_and_convert("interval_collected")?,
+            report_decode_failure: row.get_bigint_and_convert("report_decode_failure")?,
+            report_decrypt_failure: row.get_bigint_and_convert("report_decrypt_failure")?,
+            report_expired: row.get_bigint_and_convert("report_expired")?,
+            report_outdated_key: row.get_bigint_and_convert("report_outdated_key")?,
+            report_success: row.get_bigint_and_convert("report_success")?,
+            report_too_early: row.get_bigint_and_convert("report_too_early")?,
+            task_expired: row.get_bigint_and_convert("task_expired")?,
+        })
+    }
+
+    /// Add one to the counter associated with the given [`TaskId`]. The column to increment is given
+    /// by [`TaskUploadIncrementor`]. This is sharded, requiring an `ord` parameter to determine which
+    /// shard to add to. `ord` should be randomly generated by the caller.
+    #[tracing::instrument(skip(self), err)]
+    pub async fn increment_task_upload_counter(
+        &self,
+        task_id: &TaskId,
+        ord: u64,
+        incrementor: &TaskUploadIncrementor,
+    ) -> Result<(), Error> {
+        // Brute force each possible query. We cannot parameterize column names in prepared
+        // statements and we want to avoid the hazards of string interpolation into SQL.
```

Comment:
The string interpolation would be only the column name to increment, correct?

If so, I think this is a place where string interpolation to build a query makes sense. (Other, more structured, methods of building a query may be preferable, but I don't think our current dependencies allow for this & I don't want to take another dep just for this one query.)

The reason I think it is safe to build a query here: the dynamic part of the query is selected from a small set of possible values (i.e. column names based on the `TaskUploadIncrementor`), all of which can be seen as "safe" by inspection. The particular incrementor is at least partially controlled by untrusted input (i.e. an attacker can choose which kind of bad report to send), but since all possible values are known & safe by inspection, this does not matter.

Probably the best way to implement this would be to augment `TaskUploadIncrementor` to be able to provide the column name to increment, and use that to build the query.

Reply:
Fair enough, admittedly some of my hesitance to interpolate strings is paranoia. I will build the query and leave a `// safety:`-esque comment.